config/7.2.0/securebanking/ig/scripts/groovy/SignEventsResponse.groovy

```diff
@@ -0,0 +1,94 @@
+    }
+}
+
+Error response example:
+{
+    "Code": "500",
+    "Id": "ede36b552dc951d9836a127f16a7c033",
+    "Message": "[Status: 500 Internal Server Error]",
+    "Errors": [
+        {
+            "ErrorCode": "UK.OBIE.UnexpectedError",
+            "Message": "Internal error [Unknown Signing Algorithm]"
+        }
+    ]
+}
+
+Validate signature
+- You can validate the JWT SET signature in: https://jwt.davetonge.co.uk/
+- Copy a JWT 'set' value and paste in the web page.
+- JWKs URL endpoint to validate the signature: https://keystore.openbankingtest.org.uk/0015800001041REAAY/0015800001041REAAY.jwks
+ */
+
+def fapiInteractionId = request.getHeaders().getFirst("x-fapi-interaction-id");
+if (fapiInteractionId == null) fapiInteractionId = "No x-fapi-interaction-id"
+SCRIPT_NAME = "[SignEventsResponse] (" + fapiInteractionId + ") - "
+
+Map<String, Object> critClaims = Map.of(
+"http://openbanking.org.uk/iat", System.currentTimeMillis() / 1000,
+"http://openbanking.org.uk/iss", aspspOrgId,
+"http://openbanking.org.uk/tan", "openbanking.org.uk")
+
+next.handle(context, request).thenOnResult({ response ->
+    logger.debug("{} Running...", SCRIPT_NAME)
+
+    Status status = response.getStatus()
+
+    if (status.isClientError() || status.isServerError()) {
+        return newResultPromise(response)
+    }
+
+    var responseBody = response.getEntity().getJson()
+
+    Map<String, String> sets = responseBody.sets
+
+    def slurper = new JsonSlurper()
+    sets.forEach((jti, payload) -> {
+        Map payloadMap = slurper.parseText(payload)
+        signer.sign(payloadMap, critClaims)
+                .then(signedJwt -> {
+                    logger.debug("signed Jwt {}", signedJwt)
+                    responseBody.sets[jti] = signedJwt
+                }, sapiJwsSignerException -> {
+                    logger.error("{} Signature fails: {}", SCRIPT_NAME, sapiJwsSignerException.getMessage())
+                    response.status = Status.INTERNAL_SERVER_ERROR
+                    response.entity = "{ \"error\":\"" + sapiJwsSignerException.getMessage() + "\"}"
+                    return newResultPromise(response)
+                })
+    })
+
+    if (response.getStatus().isServerError()) {
+        return response
+    }
+
+    logger.debug("{} final response with signed events {}", SCRIPT_NAME, responseBody)
+    response.entity = responseBody
+    return response
```

Comment:
This needs re-writing to chain the async callbacks together to produce the Response object.

Whenever you have a Promise, you have to assume that the handler callbacks are executed by another thread and at some point in the future. The challenge with this style of programming is that the code reads sequentially but is not guaranteed to execute sequentially.

This is similar to the comment I made on the Signer Java class, this logic may currently work but that will be because the signingManager that is currently configured does not need to do any asynchronous work and the Promises it returns are are completed results. If this code does get executed in an async fashion, then the original (unsigned) response from the RS will be returned.

Added complexity in this case is that you need to do a loop, take a look at the docs which have a section on Async loops: https://community.forgerock.com/t/identity-gateway-7-1-highway-to-async-programming/28

Reply:
Refactored to use `when` Asynchronous function loop (list of promises) and return a promise (a List of results) when all promises, passed as argument, complete or throw the first exception if happens. 
Then the result `List<Map<String, String> (List<Map<jti, signedJwt>)`, or the exception is processed to build the response.